### PR TITLE
Feature/28085 docker

### PR DESCRIPTION
**Pull request by md-magenta**
Tilføjer en `Dockerfile` til udvikling og produktion samt en `docker-compose.yml` til udvikling.

Kør `docker-compose up -d --build` for at starte en postgres, rabbitmq og mox container.

Se derefter på outputtet med `docker-compose logs <service navn>`.

#28085, #28087 og #28089. ~Dokumentationen af begge kommer i et senere PR~

### CODE REVIEW

**Comment by agger-magenta** [comment]
_Skal_ dockerfilerne ligge i roden? Jeg ville foretrække, at de kom ned i en mappe for sig selv, der hed "docker" eller noget i den retning.

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,18 @@
+*.DS_Store
+.git
+.gitignore
+.gitlab-ci.yml
+Jenkinsfile
+LICENSE
```
Comment (agger-magenta) [inline .dockerignore]: Når vi distribuerer programmet, er det som udgangspunkt nødvendigt at distribuere licens-betingelserne også. 

Jeg foreslår, at vi for at sikre dette medtager licens-filen og også helst README-filen.

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    sleep 1
+done
+
+>&2 echo "Postgres is up - continuing"
+
+
+
+# Initialize the database
+# -----------------------
+# TODO This is a modified version of oio_rest/initdb.sh. This should properly be
```
Comment (agger-magenta) [inline docker-entrypoint.sh]: "properly" -> "probably"?

**Comment by md-magenta** [comment]
Dokumentation er ikke færdig endnu.

**Comment by md-magenta** [comment]
> _Skal_ dockerfilerne ligge i roden? Jeg ville foretrække, at de kom ned i en mappe for sig selv, der hed "docker" eller noget i den retning.

Det er ikke strengt nødvendigt. Det er dog kotume at de ligger der. `docker` og `docker-compose` forventer at de ligger der. Der er flag til at overstyre det eks: `docker build -f myFolder/Dockerfile` og `docker-compose -f myFolder/docker-compose.yml up -d`. Hvis de bliver flyttet skal man angive flaget og stien hver gang man skal bruge CLIen.

**Comment by md-magenta** [comment]
Jeg vil lige henvise til min commit besked f95256f0a39a432a18167a80f654535292a975b5 som motivation for fjernelsen af `mox` entry pointet.

**Comment by agger-magenta** [comment]
> > _Skal_ dockerfilerne ligge i roden? Jeg ville foretrække, at de kom ned i en mappe for sig selv, der hed "docker" eller noget i den retning.
> 
> Det er ikke strengt nødvendigt. Det er dog kotume at de ligger der. `docker` og `docker-compose` forventer at de ligger der. Der er flag til at overstyre det eks: `docker build -f myFolder/Dockerfile` og `docker-compose -f myFolder/docker-compose.yml up -d`. Hvis de bliver flyttet skal man angive flaget og stien hver gang man skal bruge CLIen.

OK, det er fint! Bare lade dem ligge.

File: doc/user/installation.rst [hunk 3]
```diff
@@ -0,0 +1,178 @@
+============
+Installation
+============
+
+All the active developed source code of mox is a python package located in
```
Comment (agger-magenta) [inline doc/user/installation.rst]: Python should be capitalized, i.e. `python` -> `Python`

File: doc/user/installation.rst [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+
+From source
+===========
+
+All the relevant code is in a python package located in :file:`oio_rest`.
+
+Prerequisites
+-------------
+
```
Comment (agger-magenta) [inline doc/user/installation.rst]: It's possible to have all of these prerequisites installed and have  a production-like setup with Postgres and RabbitMQ and gunicorn + nginx vy running the `install.sh` script. That's  simpler than performing the steps that you outline below. :slightly_smiling_face:

File: doc/user/installation.rst (hunk 3 above)
Comment (agger-magenta) [inline doc/user/installation.rst]: "active" is used adverbially here, so it should be "actively": "All the actively developed source code  ..."

File: .dockerignore [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+agentbase
+agents
+auth.sh
+compile-java.sh
+custom
+doc
+install.sh
+python_agents
+src
+venv
```
[the remark this answers is not in the document]
Comment (md-magenta) [reply .dockerignore]: I `docker-compose.yml` bliver den nuværende mappe mountet som `/code` og `CMD` bliver overskrevet så `gunicorn` genstarter på filændringer. Når du forbinder til den og kører `ls`, så ser du mappen på din host maskine. Hvis du istedet kører
```
docker build . -t mox
docker run --entrypoint ls mox
```
ser du at det kun er
```
LICENSE
README.rst
docker-entrypoint.sh
oio_rest
```
der ligger i `/code` i imaget.

File: doc/user/installation.rst (hunk 4 above)
Comment (md-magenta) [reply doc/user/installation.rst]: It is simpler _iff_  you are running Ubuntu 16.04 and want both mox, postgres, rabbitmq and nginx on the same machine, with `oio_rest` running in a virtual environment on port 8080 and connecting to postgress with user `mox` and password `mox`, nginx running on port 80 only forwarding traffic to `oio_rest`, a `oio_rest` systemd service under the user `mox` and the gunicorn logfiles written to `/var/lox/mox`.
If you want to change _any_ of those variables, you have to do it by hand. Without this documentation, you have to reverse engineer `install.sh` or `Dockerfile` to find the correct prerequisites. I instead document them here.

* If we want `install.sh` longterm, I propose to make an additional section here describing it.
* If `install.sh` is short-lived during a transition period to docker, I propose to leave the documentation as is.
* If we are comfortable enough to remove it now, I propose to remove it now.

I hope that when developers are comfortable with docker and docker-compose they don't need `install.sh` any more. Having three places where prerequisites is defined is asking for trouble. I have left in the repository as not to force anybody to use docker, but left it undocumented as not to encourage its use.

File: .dockerignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (agger-magenta) [reply .dockerignore]: Ah, OK, tak! :-)

File: doc/user/installation.rst (hunk 4 above)
Comment (agger-magenta) [reply doc/user/installation.rst]: Hmm, that's fair. I propose to leave it be for at least two release cycles, then we'll make a decision. 

I liked the idea of being able to do a complete, quick install in one fell swoop and having it close to a production-ready setup, but the Docker process is also really quick and easy, so the need for it is no longer so obvious. As I said, let's leave `installh.sh` a  bit, but if everybody seems to migrate to Docker, let's get rid of it.

**Pull request merged**